### PR TITLE
chore(STONEO11Y-99): Alerting rule for Pods Not Ready

### PR DESCRIPTION
**Pull request by Kousalya1998**
* Created a new alerting rule for pods not ready.
* Added related unit test cases.

### CODE REVIEW

**Review by amisstea — COMMENTED** [review]
I don't see any unit tests. Did you accidentally omit them?

File: prometheus/base/alerting/prometheus.pod_alerts.yaml [hunk 1]
```diff
@@ -25,3 +25,14 @@ spec:
       annotations:
         summary: "Pod {{ $labels.pod }} is crash looping"
         description: "Pod {{ $labels.namespace }}/{{ $labels.pod }} ({{ $labels.container }}) is in waiting state (reason: 'CrashLoopBackOff') for more than 15 minutes."
+    - alert: PodsNotReady
+      expr: |
+        sum by (namespace, pod)
+        (max by (namespace, pod) (kube_pod_status_phase{phase=~"Pending|Unknown|Failed", namespace!~"(.*-tenant|openshift-.*|kube-.*|default)"}
```
Comment (hmariset) [inline prometheus/base/alerting/prometheus.pod_alerts.yaml:31]: I think max by (namespace, pod) is not required here

File: test/promql/tests/test_notready_pods.yaml [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+  - interval: 1m
+    input_series:
+
+      # Pod is Pending, not unschedulable and its alerted
+      - series: 'kube_pod_status_phase{pod="pod-1", namespace="ns-1", phase="Pending"}'
+        values: '0x15'
+      - series: 'kube_pod_status_unschedulable{namespace="ns-1", pod="pod-1"}'
+        values: '0x15'
+
+      # Pod is Pending, unschedulable and its alerted
```
Comment (hmariset) [inline test/promql/tests/test_notready_pods.yaml:16]: This is not alerted. Please change the comment to not alerted.

File: test/promql/tests/test_notready_pods.yaml [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+      - series: 'kube_pod_status_unschedulable{namespace="ns-2", pod="pod-3"}'
+        values: '1x15'
+
+      # Pod is Running, not unschedulable and its not alerted
+      - series: 'kube_pod_status_phase{pod="pod-3", namespace="ns-1", phase="Running"}'
+        values: '1x10 0x5'
+      - series: 'kube_pod_status_unschedulable{namespace="ns-1", pod="pod-3"}'
+        values: '0x15'
+
+      # Pod is Pending, schedulable and its alerted
```
Comment (hmariset) [inline test/promql/tests/test_notready_pods.yaml:40]: This is also not alerted and pod is unscheduled for most of the time and only available for some window in between.

File: test/promql/tests/test_notready_pods.yaml (hunk 2 above)
Comment (Kousalya1998) [reply test/promql/tests/test_notready_pods.yaml]: > This is not alerted. Please change the comment to not alerted.

Updated

File: test/promql/tests/test_notready_pods.yaml (hunk 3 above)
Comment (Kousalya1998) [reply test/promql/tests/test_notready_pods.yaml]: Updated

File: prometheus/base/alerting/prometheus.pod_alerts.yaml [hunk 4]
```diff
@@ -25,3 +25,14 @@ spec:
...
+        unless ignoring (phase) (kube_pod_status_unschedulable{namespace!~"(.*-tenant|openshift-.*|kube-.*|default)"} == 1))) >= 0
```
Comment (yftacherzog) [inline prometheus/base/alerting/prometheus.pod_alerts.yaml:32]: Do we need to the filtering here? i.e. the `{namespace!~"(.*-tenant|openshift-.*|kube-.*|default)"}` part.
The `unless` operator already suppose to filter out all entries which does not have the same values for all labels. If we can drop this, it will simplify the rule and make it more maintainable.

File: prometheus/base/alerting/prometheus.pod_alerts.yaml [hunk 5]
```diff
@@ -25,3 +25,14 @@ spec:
       annotations:
         summary: "Pod {{ $labels.pod }} is crash looping"
         description: "Pod {{ $labels.namespace }}/{{ $labels.pod }} ({{ $labels.container }}) is in waiting state (reason: 'CrashLoopBackOff') for more than 15 minutes."
+    - alert: PodsNotReady
+      expr: |
```
Comment (yftacherzog) [inline prometheus/base/alerting/prometheus.pod_alerts.yaml:29]: I don't think the expression is correct.

As far as I understand it, all pods have entries for all phases all the time and the value is `1` for the entry with the current phase. I think we need to verify that the value is `== 1` rather than `>= 0`.

I think the current rule will alert constantly. You can verify this by adding entries for all phases on a unit test.

e.g.

```
      - series: 'kube_pod_status_phase{pod="pod-1", namespace="ns-1", phase="Pending"}'
        values: '0x15'
      - series: 'kube_pod_status_phase{pod="pod-1", namespace="ns-1", phase="Unknown"}'
        values: '0x15'
      - series: 'kube_pod_status_phase{pod="pod-1", namespace="ns-1", phase="Failed"}'
        values: '0x15'
      - series: 'kube_pod_status_phase{pod="pod-1", namespace="ns-1", phase="Running"}'
        values: '1x15'
      - series: 'kube_pod_status_phase{pod="pod-1", namespace="ns-1", phase="Succeeded"}'
        values: '0x15'
      - series: 'kube_pod_status_unschedulable{namespace="ns-1", pod="pod-1"}'
        values: '0x15'
```

File: prometheus/base/alerting/prometheus.pod_alerts.yaml [hunk 6]
```diff
@@ -25,3 +25,14 @@ spec:
...
+        sum by (namespace, pod)
```
[the remark this answers is not in the document]
Comment (yftacherzog) [reply prometheus/base/alerting/prometheus.pod_alerts.yaml:30]: I think it should be something like this (not tested):
```
        kube_pod_status_phase{phase=~"Pending|Unknown|Failed", namespace!~"(.*-tenant|openshift-.*|kube-.*|default)"} == 1
        unless ignoring (phase) (kube_pod_status_unschedulable == 1)
```

File: test/promql/tests/test_notready_pods.yaml [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+
+rule_files:
+  - ../extracted-rules.yaml
+
+tests:
+  - interval: 1m
+    input_series:
+
+      # Pod is Pending, not unschedulable and its alerted
+      - series: 'kube_pod_status_phase{pod="pod-1", namespace="ns-1", phase="Pending"}'
```
Comment (yftacherzog) [inline test/promql/tests/test_notready_pods.yaml:11]: We need all phases for all pods - at least for some of the test cases.

File: test/promql/tests/test_notready_pods.yaml [hunk 8]
```diff
@@ -0,0 +1,88 @@
+evaluation_interval: 1m
+
+rule_files:
+  - ../extracted-rules.yaml
+
+tests:
+  - interval: 1m
+    input_series:
+
+      # Pod is Pending, not unschedulable and its alerted
```
Comment (yftacherzog) [inline test/promql/tests/test_notready_pods.yaml:10]:
```suggestion
      # Pod is Pending, not unschedulable and it's alerted
```

File: test/promql/tests/test_notready_pods.yaml [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+      # Pod is Failed, Unscheduled the whole time but the namespace is default so it's ignored
+      - series: 'kube_pod_status_phase{pod="pod-6", namespace="default", phase="Failed"}'
+        values: '0x15'
+      - series: 'kube_pod_status_unschedulable{namespace="default", pod="pod-6"}'
+        values: '1x15'
+    alert_rule_test:
+      - eval_time: 15m
+        alertname: PodsNotReady
+        exp_alerts:
+          # Alert 1.
```
Comment (yftacherzog) [inline test/promql/tests/test_notready_pods.yaml:73]: Delete comment or replace with something non-generic

**Review by yftacherzog — CHANGES_REQUESTED** [review]
I don't think the provided rule is correct. See my comments for details.

File: prometheus/base/alerting/prometheus.pod_alerts.yaml (hunk 1 above)
Comment (Kousalya1998) [reply prometheus/base/alerting/prometheus.pod_alerts.yaml]: removed

File: prometheus/base/alerting/prometheus.pod_alerts.yaml (hunk 4 above)
Comment (Kousalya1998) [reply prometheus/base/alerting/prometheus.pod_alerts.yaml]: I've updated the expression now.

File: prometheus/base/alerting/prometheus.pod_alerts.yaml (hunk 5 above)
Comment (Kousalya1998) [reply prometheus/base/alerting/prometheus.pod_alerts.yaml]: I've updated the test case file according to the new expression.

File: prometheus/base/alerting/prometheus.pod_alerts.yaml (hunk 6 above)
[the remark this answers is not in the document]
Comment (Kousalya1998) [reply prometheus/base/alerting/prometheus.pod_alerts.yaml]: I've updated the expression now.

File: test/promql/tests/test_notready_pods.yaml (hunk 9 above)
Comment (Kousalya1998) [reply test/promql/tests/test_notready_pods.yaml]: Deleted.

File: test/promql/tests/test_notready_pods.yaml (hunk 7 above)
Comment (Kousalya1998) [reply test/promql/tests/test_notready_pods.yaml:11]: I've updated the unit test cases now.

File: test/promql/tests/test_notready_pods.yaml [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+      - series: 'kube_pod_status_unschedulable{namespace="ns-3", pod="pod-1"}'
+        values: '1x1 0x14'
+
+      # Pod is in the Running state and scheduled, so it will not be alerted.
+      - series: 'kube_pod_status_phase{pod="pod-2", namespace="ns-1", phase="Running"}'
+        values: '1x15'
+      - series: 'kube_pod_status_unschedulable{namespace="ns-1", pod="pod-2"}'
+        values: '0x15'
+
+      # Pod is in the Pending state and scheduled, but it has a namespace that starts with 'openshift' so its ignored.
```
Comment (yftacherzog) [inline test/promql/tests/test_notready_pods.yaml:44]: typo: its -> it's

File: test/promql/tests/test_notready_pods.yaml [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+      - series: 'kube_pod_status_phase{pod="pod-2", namespace="openshift-prod-test", phase="Pending"}'
+        values: '1x15'
+      - series: 'kube_pod_status_unschedulable{namespace="openshift-prod-test", pod="pod-2"}'
+        values: '0x15'
+
+      # Pod is in the Failed state and scheduled, but it has a namespace that ends with 'tenant' so its ignored.
```
Comment (yftacherzog) [inline test/promql/tests/test_notready_pods.yaml:50]: typo: its -> it's

File: test/promql/tests/test_notready_pods.yaml [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+      - series: 'kube_pod_status_phase{pod="pod-1", namespace="ns-1", phase="Failed"}'
+        values: '0x15'
+      - series: 'kube_pod_status_phase{pod="pod-1", namespace="ns-1", phase="Running"}'
+        values: '0x15'
+      - series: 'kube_pod_status_phase{pod="pod-1", namespace="ns-1", phase="Succeeded"}'
+        values: '0x15'
+      - series: 'kube_pod_status_unschedulable{namespace="ns-1", pod="pod-1"}'
+        values: '0x15'
+
+      # Pod is in the Unknown state and unscheduled(unscheduled once in the 15 min interval), so it will not be alerted.
```
Comment (yftacherzog) [inline test/promql/tests/test_notready_pods.yaml:24]: Nice test case!

File: test/promql/tests/test_notready_pods.yaml [hunk 13]
```diff
@@ -0,0 +1,67 @@
+evaluation_interval: 1m
+
+rule_files:
+  - ../extracted-rules.yaml
+
+tests:
+  - interval: 1m
```
Comment (yftacherzog) [inline test/promql/tests/test_notready_pods.yaml:7]: We only have 5 phases, so I think we can add the "regular" case for each. i.e. when the pod is not unschedulable, verify that each of the phases acts as expected - alerting for the 3 bad phases and not alerting for the good ones.

File: test/promql/tests/test_notready_pods.yaml [hunk 14]
```diff
@@ -0,0 +1,67 @@
+evaluation_interval: 1m
+
+rule_files:
+  - ../extracted-rules.yaml
+
```
Comment (yftacherzog) [inline test/promql/tests/test_notready_pods.yaml:5]: For readability and maintainability, I suggest making the test cases independent. i.e. have each of them as a separate instance of the `tests` list. You can have all the non-alerting ones in one test and each alerting in a test of its own. Something like this:

```
tests:
  # Pod is in the Pending state and scheduled, so it will be alerted.
  - interval: 1m
    input_series:

      - series: 'kube_pod_status_phase{pod="pod-1", namespace="ns-1", phase="Pending"}'
        values: '1x15'
      - series: 'kube_pod_status_phase{pod="pod-1", namespace="ns-1", phase="Unknown"}'
        values: '0x15'
      - series: 'kube_pod_status_phase{pod="pod-1", namespace="ns-1", phase="Failed"}'
        values: '0x15'
      - series: 'kube_pod_status_phase{pod="pod-1", namespace="ns-1", phase="Running"}'
        values: '0x15'
      - series: 'kube_pod_status_phase{pod="pod-1", namespace="ns-1", phase="Succeeded"}'
        values: '0x15'
      - series: 'kube_pod_status_unschedulable{namespace="ns-1", pod="pod-1"}'
        values: '0x15'

    alert_rule_test:
      - eval_time: 15m
        alertname: PodsNotReady
        exp_alerts:
          - exp_labels:
              severity: warning
              namespace: ns-1
              pod: pod-1
              phase: Pending
            exp_annotations:
              summary: "Pod pod-1 is not ready"
              description: "Pod pod-1 in namespace ns-1 is not ready for more than 15 minutes."


  # Pod is in the Unknown state and scheduled, so it will be alerted.
  - interval: 1m
    input_series:

      - series: 'kube_pod_status_phase{pod="pod-1", namespace="ns-1", phase="Pending"}'
        values: '0x15'
      - series: 'kube_pod_status_phase{pod="pod-1", namespace="ns-1", phase="Unknown"}'
        values: '1x15'
      - series: 'kube_pod_status_phase{pod="pod-1", namespace="ns-1", phase="Failed"}'
        values: '0x15'
      - series: 'kube_pod_status_phase{pod="pod-1", namespace="ns-1", phase="Running"}'
        values: '0x15'
      - series: 'kube_pod_status_phase{pod="pod-1", namespace="ns-1", phase="Succeeded"}'
        values: '0x15'
      - series: 'kube_pod_status_unschedulable{namespace="ns-1", pod="pod-1"}'
        values: '0x15'

    alert_rule_test:
      - eval_time: 15m
        alertname: PodsNotReady
        exp_alerts:
          - exp_labels:
              severity: warning
              namespace: ns-1
              pod: pod-1
              phase: Pending
            exp_annotations:
              summary: "Pod pod-1 is not ready"
              description: "Pod pod-1 in namespace ns-1 is not ready for more than 15 minutes."

  # Some other test case
  - interval: 1m
    input_series:
    ...
```

File: test/promql/tests/test_notready_pods.yaml
Comment (yftacherzog) [inline test/promql/tests/test_notready_pods.yaml:1]: Can you also add a test case in which the phase is cycling between the alerting phases within a period of 15 minutes? I think it should alert as well.

Something like this:

```
      - series: 'kube_pod_status_phase{pod="pod-1", namespace="ns-1", phase="Pending"}'
        values: '1x4 0x4 0x4'
      - series: 'kube_pod_status_phase{pod="pod-1", namespace="ns-1", phase="Unknown"}'
        values: '0x4 1x4 0x4'
      - series: 'kube_pod_status_phase{pod="pod-1", namespace="ns-1", phase="Failed"}'
        values: '0x4 0x4 1x4'
      - series: 'kube_pod_status_phase{pod="pod-1", namespace="ns-1", phase="Running"}'
        values: '0x15'
      - series: 'kube_pod_status_phase{pod="pod-1", namespace="ns-1", phase="Succeeded"}'
        values: '0x15'
      - series: 'kube_pod_status_unschedulable{namespace="ns-1", pod="pod-1"}'
        values: '0x15'
```

File: test/promql/tests/test_notready_pods.yaml [hunk 15]
```diff
@@ -0,0 +1,67 @@
...
+        values: '0x15'
+      - series: 'kube_pod_status_phase{pod="pod-1", namespace="ns-3", phase="Running"}'
+        values: '0x15'
+      - series: 'kube_pod_status_phase{pod="pod-1", namespace="ns-3", phase="Succeeded"}'
+        values: '0x15'
+      - series: 'kube_pod_status_unschedulable{namespace="ns-3", pod="pod-1"}'
+        values: '1x1 0x14'
+
+      # Pod is in the Running state and scheduled, so it will not be alerted.
+      - series: 'kube_pod_status_phase{pod="pod-2", namespace="ns-1", phase="Running"}'
```
Comment (yftacherzog) [inline test/promql/tests/test_notready_pods.yaml:39]: Your code is not actually detecting the `Running` phase so I think you're missing the part of testing it. You need to also include the values of the metrics that your expression is checking. i.e. include the lines for all the bad phases and give them value of `0` throughout.

File: test/promql/tests/test_notready_pods.yaml (hunk 12 above)
Comment (yftacherzog) [reply test/promql/tests/test_notready_pods.yaml:24]: Also add a test case in which the pod is not unschedulable but it's within the bad phase for 14 minutes rather than 15 minutes, and in a good phase for the other minute.

Something like this:
```
      - series: 'kube_pod_status_phase{pod="pod-1", namespace="ns-1", phase="Failed"}'
        values: '0 1x13'
      - series: 'kube_pod_status_phase{pod="pod-1", namespace="ns-1", phase="Running"}'
        values: '1 0x13'
```

In this case, as far as I see it, you don't have to include all the different phases - just one good and one bad.

File: test/promql/tests/test_notready_pods.yaml
Comment (yftacherzog) [reply test/promql/tests/test_notready_pods.yaml:1]: You can see a reference for the `axb` notation here: https://prometheus.io/docs/prometheus/latest/configuration/unit_testing_rules/#series

File: test/promql/tests/test_notready_pods.yaml [hunk 16]
```diff
@@ -0,0 +1,67 @@
...
+      # Pod is in the Pending state and scheduled, but it has a namespace that starts with 'openshift' so its ignored.
+      - series: 'kube_pod_status_phase{pod="pod-2", namespace="openshift-prod-test", phase="Pending"}'
+        values: '1x15'
+      - series: 'kube_pod_status_unschedulable{namespace="openshift-prod-test", pod="pod-2"}'
+        values: '0x15'
+
+      # Pod is in the Failed state and scheduled, but it has a namespace that ends with 'tenant' so its ignored.
+      - series: 'kube_pod_status_phase{pod="pod-1", namespace="kube-tenant", phase="Failed"}'
+        values: '1x15'
+      - series: 'kube_pod_status_unschedulable{namespace="kube-test", pod="pod-1"}'
```
Comment (hmariset) [inline test/promql/tests/test_notready_pods.yaml:53]: namespace for both the metrics are not matching, change either of them

Comment (Kousalya1998) [reply test/promql/tests/test_notready_pods.yaml:53]: > namespace for both the metrics are not matching, change either of them

Ty, updated.

File: test/promql/tests/test_notready_pods.yaml
Comment (Kousalya1998) [reply test/promql/tests/test_notready_pods.yaml:1]: @yftacherzog is there a reason why you are giving it like `1x4 0x4 0xx4` instead of `1x5 0x5 0x5`?

Comment (yftacherzog) [reply test/promql/tests/test_notready_pods.yaml:1]: Yes. Have a look at the link I shared. It's explained there.
```
#     3. ' 1x4' becomes '1 1 1 1 1' - shorthand for '1+0x4', series starts at 1, then 4 further samples incrementing by 0.
```

Comment (Kousalya1998) [reply test/promql/tests/test_notready_pods.yaml:1]: @yftacherzog I did take a look at the reference document. I think that line is wrong. 

Based on my research online, I found the following:

```
For example, the string '-2+4x3' specifies a series that starts at -2 and increments by 4 to produce the sequence -2, 2, 6, and 10. The string '1-2x4' specifies a series that starts at 1 and decrements by 2 to produce the sequence 1, -1, -3, -5, and -7. The string '1x4' specifies a series that starts at 1 and does not change over 4 samples, resulting in the sequence 1, 1, 1, and 1. 
```
My assumptions:
`1x4` will have a sequence `1, 1, 1, 1`
and
`1x5` will have a sequence `1, 1, 1, 1, 1`


Not sure if the below line is correct in the reference document?
`#     3. ' 1x4' becomes '1 1 1 1 1' - shorthand for '1+0x4', series starts at 1, then 4 further samples incrementing by 0.`

File: test/promql/tests/test_notready_pods.yaml [hunk 17]
```diff
@@ -0,0 +1,162 @@
...
+      # - series: 'kube_pod_status_phase{pod="pod-1", namespace="ns-6", phase="Unknown"}'
+      #   values: '0x4 1x4 0x4'
+      # - series: 'kube_pod_status_phase{pod="pod-1", namespace="ns-6", phase="Failed"}'
+      #   values: '0x4 0x4 1x4'
+      # - series: 'kube_pod_status_phase{pod="pod-1", namespace="ns-6", phase="Running"}'
+      #   values: '0x14'
+      # - series: 'kube_pod_status_phase{pod="pod-1", namespace="ns-6", phase="Succeeded"}'
+      #   values: '0x14'
+      # - series: 'kube_pod_status_unschedulable{namespace="ns-6", pod="pod-1"}'
+      #   values: '0x14'
```
Comment (yftacherzog) [inline test/promql/tests/test_notready_pods.yaml:138]: Remove commented-out code

Comment (yftacherzog) [inline test/promql/tests/test_notready_pods.yaml:138]: Why is this commented out? This should actually be alerted. No?

File: test/promql/tests/test_notready_pods.yaml [hunk 18]
```diff
@@ -0,0 +1,162 @@
...
+      - series: 'kube_pod_status_phase{pod="pod-1", namespace="ns-5", phase="Pending"}'
+        values: '0x14'
+      - series: 'kube_pod_status_phase{pod="pod-1", namespace="ns-5", phase="Unknown"}'
+        values: '0x14'
+      - series: 'kube_pod_status_phase{pod="pod-1", namespace="ns-5", phase="Failed"}'
+        values: '0x14'
+      - series: 'kube_pod_status_phase{pod="pod-1", namespace="ns-5", phase="Running"}'
+        values: '1x14'
+      - series: 'kube_pod_status_phase{pod="pod-1", namespace="ns-5", phase="Succeeded"}'
+        values: '1x14'
```
Comment (yftacherzog) [inline test/promql/tests/test_notready_pods.yaml:122]: It's Running and Succeeded at the same time. I don't think this is valid.

**Review by yftacherzog — CHANGES_REQUESTED** [review]
Almost there. Good progress!

File: test/promql/tests/test_notready_pods.yaml (hunk 18 above)
Comment (Kousalya1998) [reply test/promql/tests/test_notready_pods.yaml:122]: Ty, I updated this line.

**Comment by Kousalya1998** [comment]
@yftacherzog @hmariset @avi-biton Does anyone know why some checks haven't been completed for this PR? Is it showing half success and half still checking?

**Comment by yftacherzog** [comment]
> @yftacherzog @hmariset @avi-biton Does anyone know why some checks haven't been completed for this PR? it is showing half success and half still checking.

Not sure. There's some sort of leftover legacy app that was there as well. It might be that this is the one that did not complete. In that case, it should not block the merge. Our tests seem to be working as far as I can tell.

File: test/promql/tests/test_notready_pods.yaml [hunk 19]
```diff
@@ -0,0 +1,162 @@
...
+      - series: 'kube_pod_status_phase{pod="pod-1", namespace="ns-5", phase="Failed"}'
+        values: '0x14'
+      - series: 'kube_pod_status_phase{pod="pod-1", namespace="ns-5", phase="Running"}'
+        values: '0x14'
+      - series: 'kube_pod_status_phase{pod="pod-1", namespace="ns-5", phase="Succeeded"}'
+        values: '1x14'
+      - series: 'kube_pod_status_unschedulable{namespace="ns-5", pod="pod-1"}'
+        values: '0x14'
+
+      # For the first 5 minutes, Pod is in the Pending state, next 5 minutes it's in the Unknown state, for last 5 minutes it's in the Failed state and scheduled, so it will be not alerted.
```
Comment (yftacherzog) [inline test/promql/tests/test_notready_pods.yaml:126]: I'd expect it to be alerting. Do you know why it is not alerting? Judging by the rule itself, it should be alerting. Might be that there's some sort of edge case here.

Comment (yftacherzog) [reply test/promql/tests/test_notready_pods.yaml:126]: Maybe try with something like this to cover some sort of edge case with the transition between the stages:
```
      - series: 'kube_pod_status_phase{pod="pod-1", namespace="ns-6", phase="Pending"}'
        values: '1x5 0x3 0x4'
      - series: 'kube_pod_status_phase{pod="pod-1", namespace="ns-6", phase="Unknown"}'
        values: '0x4 1x5 0x3'
      - series: 'kube_pod_status_phase{pod="pod-1", namespace="ns-6", phase="Failed"}'
        values: '0x4 0x4 1x4'
```
I would not expect that to be an issue, but I think this sort of setup should rule that out at least.

File: prometheus/base/alerting/prometheus.pod_alerts.yaml [hunk 20]
```diff
@@ -25,3 +25,11 @@ spec:
       annotations:
         summary: "Pod {{ $labels.pod }} is crash looping"
         description: "Pod {{ $labels.namespace }}/{{ $labels.pod }} ({{ $labels.container }}) is in waiting state (reason: 'CrashLoopBackOff') for more than 15 minutes."
+    - alert: PodsNotReady
+      expr: kube_pod_status_phase{phase=~"Pending|Unknown|Failed", namespace!~"(.*-tenant|openshift-.*|kube-.*|default)"} == 1 unless ignoring (phase) (kube_pod_status_unschedulable == 1)
```
Comment (yftacherzog) [inline prometheus/base/alerting/prometheus.pod_alerts.yaml:29]: Can we break this line a bit for readability? Something like this:
```suggestion
      expr: |
            kube_pod_status_phase{phase=~"Pending|Unknown|Failed", namespace!~"(.*-tenant|openshift-.*|kube-.*|default)"} == 1
            unless ignoring (phase) (kube_pod_status_unschedulable == 1)
```

File: test/promql/tests/test_notready_pods.yaml [hunk 21]
```diff
@@ -0,0 +1,162 @@
...
+      - series: 'kube_pod_status_phase{pod="pod-1", namespace="ns-6", phase="Failed"}'
+        values: '0x4 0x4 1x4'
+      - series: 'kube_pod_status_phase{pod="pod-1", namespace="ns-6", phase="Running"}'
+        values: '0x14'
+      - series: 'kube_pod_status_phase{pod="pod-1", namespace="ns-6", phase="Succeeded"}'
+        values: '0x14'
+      - series: 'kube_pod_status_unschedulable{namespace="ns-6", pod="pod-1"}'
+        values: '0x14'
+
+      # Pod is in the Running state for the first 14 min, then it's in the Failed state for the last 1 min and unscheduled, so it will not be alerted.
```
Comment (yftacherzog) [inline test/promql/tests/test_notready_pods.yaml:140]:
```suggestion
      # Pod is in the Running state for the first 14 min, then it's in the Failed state
      # for the last 1 min and unscheduled, so it will not be alerted.
```

File: test/promql/tests/test_notready_pods.yaml [hunk 22]
```diff
@@ -0,0 +1,162 @@
...
+              pod: pod-1
+              phase: Failed
+            exp_annotations:
+              summary: "Pod pod-1 is not ready"
+              description: "Pod pod-1 in namespace ns-3 is not ready for more than 15 minutes."
+
+  - interval: 1m
+    input_series:
+      # Not Alerted cases:
+      # Pod is in the Unknown state and unscheduled (unscheduled once in the 15 min interval), so it will not be alerted.
```
Comment (yftacherzog) [inline test/promql/tests/test_notready_pods.yaml:98]:
```suggestion
      # Pod is in the Unknown state and unscheduled (unscheduled once in the 15 min
      # interval), so it will not be alerted.
```

File: test/promql/tests/test_notready_pods.yaml [hunk 23]
```diff
@@ -0,0 +1,162 @@
...
+
+      # Pod is in the Running state for the first 14 min, then it's in the Failed state for the last 1 min and unscheduled, so it will not be alerted.
+      - series: 'kube_pod_status_phase{pod="pod-1", namespace="ns-7", phase="Running"}'
+        values: '1x13 0'
+      - series: 'kube_pod_status_phase{pod="pod-1", namespace="ns-7", phase="Failed"}'
+        values: '0x13 1'
+      - series: 'kube_pod_status_unschedulable{namespace="ns-7", pod="pod-1"}'
+        values: '0x14'
+
+      # Pod is in the Pending state and scheduled, but it has a namespace that starts with 'openshift' so it's ignored.
```
Comment (yftacherzog) [inline test/promql/tests/test_notready_pods.yaml:148]:
```suggestion
      # Pod is in the Pending state and scheduled, but it has a namespace that starts
      # with 'openshift' so it's ignored.
```

File: test/promql/tests/test_notready_pods.yaml [hunk 24]
```diff
@@ -0,0 +1,162 @@
...
+      - series: 'kube_pod_status_phase{pod="pod-2", namespace="openshift-prod-test", phase="Pending"}'
+        values: '1x15'
+      - series: 'kube_pod_status_unschedulable{namespace="openshift-prod-test", pod="pod-2"}'
+        values: '0x15'
+
+      # Pod is in the Failed state and scheduled, but it has a namespace that ends with 'tenant' so it's ignored.
```
Comment (yftacherzog) [inline test/promql/tests/test_notready_pods.yaml:154]:
```suggestion
      # Pod is in the Failed state and scheduled, but it has a namespace that ends with
      # 'tenant' so it's ignored.
```

**Comment by Kousalya1998** [comment]
> > @yftacherzog @hmariset @avi-biton Does anyone know why some checks haven't been completed for this PR? it is showing half success and half still checking.
> 
> Not sure. There's some sort of leftover legacy app that was there as well. It might be that this is the one that did not complete. In that case, it should not block the merge. Our tests seem to be working as far as I can tell.

Understood. Ty

**Review by hmariset — APPROVED** [review]
LGTM

File: test/promql/tests/test_notready_pods.yaml [hunk 25]
```diff
@@ -0,0 +1,153 @@
...
+      # Pod is in the Pending state and scheduled, but it has a namespace that starts
+      # with 'openshift' so it's ignored.
+      - series: 'kube_pod_status_phase{pod="pod-2", namespace="openshift-prod-test", phase="Pending"}'
+        values: '1x15'
+      - series: 'kube_pod_status_unschedulable{namespace="openshift-prod-test", pod="pod-2"}'
+        values: '0x15'
+
+      # Pod is in the Failed state and scheduled, but it has a namespace that ends
+      # with 'tenant' so it's ignored.
+      - series: 'kube_pod_status_phase{pod="pod-1", namespace="kube-tenant", phase="Failed"}'
```
Comment (yftacherzog) [inline test/promql/tests/test_notready_pods.yaml:146]: This one both starts with kube and ends with tenant so we can't tell which one is it not alerting for.
Let's change it to something that does not start with kube, and add additional test cases to cover the rest of that regex (i.e. one with kube and one with default)

Comment (Kousalya1998) [reply test/promql/tests/test_notready_pods.yaml]: Added all the cases.

File: test/promql/tests/test_notready_pods.yaml [hunk 26]
```diff
@@ -0,0 +1,167 @@
...
+
+      # Pod is in the Pending state and scheduled, but it has a namespace that starts
+      # with 'kube' so it's ignored.
+      - series: 'kube_pod_status_phase{pod="pod-3", namespace="kube-test", phase="Unknown"}'
+        values: '1x14'
+      - series: 'kube_pod_status_unschedulable{namespace="kube-test", pod="pod-3"}'
+        values: '1x14'
+
+      # Pod is in the Pending state and scheduled, but it has a namespace that starts
+      # with 'default' so it's ignored.
```
Comment (yftacherzog) [inline test/promql/tests/test_notready_pods.yaml:159]:
```suggestion
      # Pod is in the Pending state and scheduled, but it has namespace 'default'
      # so it's ignored.
```

File: test/promql/tests/test_notready_pods.yaml [hunk 27]
```diff
@@ -0,0 +1,167 @@
...
+        values: '0x14'
+
+      # Pod is in the Pending state and scheduled, but it has a namespace that starts
+      # with 'kube' so it's ignored.
+      - series: 'kube_pod_status_phase{pod="pod-3", namespace="kube-test", phase="Unknown"}'
+        values: '1x14'
+      - series: 'kube_pod_status_unschedulable{namespace="kube-test", pod="pod-3"}'
+        values: '0x14'
+
+      # Pod is in the Pending state and scheduled, but it has a namespace 'default
```
Comment (yftacherzog) [inline test/promql/tests/test_notready_pods.yaml:158]:
```suggestion
      # Pod is in the Pending state and scheduled, but it has a namespace 'default'
```

**Review by yftacherzog — APPROVED** [review]

File: test/promql/tests/test_notready_pods.yaml (hunk 27 above)
Comment (Kousalya1998) [reply test/promql/tests/test_notready_pods.yaml]: Thank you for correcting all my tiny errors @yftacherzog

**Comment by yftacherzog** [comment]
@gbenhaim Do you know what might be the issue here with the additional required test?

**Review by Omeramsc — APPROVED** [review]
LGTM

**Review by avi-biton — APPROVED** [review]
LGTM
